src/kmeans.jl

```diff
@@ -238,49 +236,44 @@ end
 #  (specific to the case where samples are not weighted)
 #
 function update_centers!(
-    x::Matrix{T},                   # in: sample matrix (d x n)
-    w::Nothing,                        # in: sample weights
-    assignments::Vector{Int},       # in: assignments (n)
-    to_update::Vector{Bool},        # in: whether a center needs update (k)
-    centers::Matrix{T},             # out: updated centers (d x k)
-    cweights::Vector) where T<:AbstractFloat               # out: updated cluster weights (k)
+    X::AbstractMatrix{<:Real},       # in: sample matrix (d x n)
+    w::Nothing,                      # in: sample weights
+    assignments::Vector{Int},        # in: assignments (n)
+    to_update::Vector{Bool},         # in: whether a center needs update (k)
+    centers::AbstractMatrix{<:Real}, # out: updated centers (d x k)
+    cweights::Vector{Float64})       # out: updated cluster weights (k)
 
-    d::Int = size(x, 1)
-    n::Int = size(x, 2)
-    k::Int = size(centers, 2)
+    (d, n), k = size(X), size(centers, 2)
 
     # initialize center weights
-    for i = 1 : k
-        if to_update[i]
-            cweights[i] = 0.
-        end
-    end
+    cweights[to_update] .= 0.0
 
     # accumulate columns
-    @inbounds for j = 1 : n
+    @inbounds for j = 1:n
         cj = assignments[j]
         1 <= cj <= k || error("assignment out of boundary.")
+
         if to_update[cj]
             if cweights[cj] > 0
                 for i = 1:d
```

Comment:
While you're on it, and since you're already doing these kind of changes (don't know whether it's possible to do multi-line suggestions)
```suggestion
                centers[:, cj] .+= view(X, :, j)
```